File: api/GitSrv.Api/Identity/Slug.cs
using System.Text.RegularExpressions;

namespace GitSrv.Api.Identity;

/// <summary>
/// Slug validation and normalisation for usernames, org, team and repo identifiers. Slugs are
/// stored already-normalised, so uniqueness is a plain UNIQUE constraint.
/// </summary>
public static partial class Slug
{
    [GeneratedRegex(@"^[a-z0-9](?:[a-z0-9]|[-_](?![-_]))*[a-z0-9]$|^[a-z0-9]$")]
    private static partial Regex Pattern();

    public const int MaxLength = 40;

    // Names that would collide with top-level routes or have special meaning.
    private static readonly HashSet<string> Reserved = new(StringComparer.Ordinal)
    {
        "admin", "api", "assets", "css", "js", "static", "public", "health", "login", "logout",
        "signin", "signup", "register", "settings", "help", "about", "new", "explore", "search",
        "notifications", "dashboard", "orgs", "organizations", "organisations", "teams", "users",
        "user", "repo", "repos", "repositories", "git", "gitsrv", "www", "mail", "ssh", "-", "_",
        "abuse", "security", "status", "billing", "pricing", "docs", "blog", "support",
    };

    public static bool IsValid(string candidate) =>
        !string.IsNullOrEmpty(candidate)
        && candidate.Length <= MaxLength
        && Pattern().IsMatch(candidate)
        && !Reserved.Contains(candidate);

    public static bool IsReserved(string candidate) => Reserved.Contains(candidate.ToLowerInvariant());

    /// <summary>Lowercase and trim; does not guarantee validity — call <see cref="IsValid"/> after.</summary>
    public static string Normalise(string input) => input.Trim().ToLowerInvariant();

    /// <summary>Best-effort slug from a display name, for pre-filling forms.</summary>
    public static string Suggest(string name)
    {
        var lower = name.Trim().ToLowerInvariant();
        var slug = SlugifyPattern().Replace(lower, "-").Trim('-', '_');
        if (slug.Length > MaxLength)
            slug = slug[..MaxLength].Trim('-', '_');
        return slug;
    }

    [GeneratedRegex(@"[^a-z0-9]+")]
    private static partial Regex SlugifyPattern();
}

File: web/src/js/features/slug.js
import { api } from '../api.js';

// Mirrors the server's Slug.Suggest: lowercase, non-alphanumerics -> '-', collapse, trim, cap 40.
export function suggestSlug(name) {
  return (name || '')
    .trim().toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^[-_]+|[-_]+$/g, '')
    .slice(0, 40)
    .replace(/^[-_]+|[-_]+$/g, '');
}

const RESERVED = new Set([
  'admin', 'api', 'settings', 'help', 'about', 'new', 'explore', 'search', 'login', 'logout',
  'signin', 'signup', 'register', 'orgs', 'teams', 'users', 'repos', 'git', 'gitsrv', 'www',
]);
const VALID = /^[a-z0-9](?:[a-z0-9]|[-_](?![-_]))*[a-z0-9]$|^[a-z0-9]$/;

/** Client-side shape check before we bother the server. */
export function localSlugError(slug) {
  if (!slug) return 'Required.';
  if (slug.length > 40) return 'Too long (40 max).';
  if (RESERVED.has(slug)) return `"${slug}" is reserved.`;
  if (!VALID.test(slug)) return 'Lowercase letters, digits, single - or _ between.';
  return null;
}

// check(value) implementations for form fields --------------------------

export function orgSlugCheck() {
  return async (slug) => {
    const local = localSlugError(slug);
    if (local) return { ok: false, message: local };
    const r = await api.get(`/api/slug-available/org?slug=${encodeURIComponent(slug)}`);
    return { ok: r.available, message: r.available ? 'Available' : 'That org slug is taken.' };
  };
}

export function repoSlugCheck(orgSlug) {
  return async (slug) => {
    const local = localSlugError(slug);
    if (local) return { ok: false, message: local };
    const r = await api.get(`/api/orgs/${orgSlug}/slug-available?kind=repo&slug=${encodeURIComponent(slug)}`);
    return { ok: r.available, message: r.available ? 'Available' : 'That repo slug already exists in this org.' };
  };
}

export function teamSlugCheck(orgSlug) {
  return async (slug) => {
    const local = localSlugError(slug);
    if (local) return { ok: false, message: local };
    const r = await api.get(`/api/orgs/${orgSlug}/slug-available?kind=team&slug=${encodeURIComponent(slug)}`);
    return { ok: r.available, message: r.available ? 'Available' : 'That team slug already exists in this org.' };
  };
}
